### PR TITLE
Added CloudFront Functions example

### PR DESCRIPTION
**Pull request by paulosergio-jnr**
*Issue:* https://github.com/aws-cloudformation/iac-model-evaluation/issues/12

*Description of changes:*
Added CloudFront Functions example with both request and response functions.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+      code: FunctionCode.fromFile({
+        filePath: './cloudfront-functions/response-function.js'
+      })
+    });
+
+    const cloudfrontOAI = new OriginAccessIdentity(this, 'cloudfront-OAI', {
+      comment: `OAI for ${id}`
+    });
+
+    bucket.addToResourcePolicy(new PolicyStatement({
```
Comment (ericzbeard) [inline cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts:51]: Make sure to comment every code block. And add a top level comment to the file that describes the whole thing.

File: cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    // create a cloudFront function from the response-function.js file
+    const responseFunction = new Function(this, 'ResponseFunction', {
+      functionName: 'ResponseFunction',
+      runtime: FunctionRuntime.JS_2_0,
+      code: FunctionCode.fromFile({
+        filePath: './cloudfront-functions/response-function.js'
+      })
+    });
+
+    const cloudfrontOAI = new OriginAccessIdentity(this, 'cloudfront-OAI', {
```
Comment (ericzbeard) [inline cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts:47]: Can we use an OAC instead since that is best practice?

**Review by ericzbeard — CHANGES_REQUESTED** [review]

**Review by ericzbeard — COMMENTED** [review]
Why did we go with cloudfront-functions as the name? Why not expand on the existing cloudfront sample?

**Comment by paulosergio-jnr** [comment]
> Why did we go with cloudfront-functions as the name? Why not expand on the existing cloudfront sample?

It was just because the `-headers` suffix seemed to be something very specific.

WDYT of merging both examples, applying the comments above, and changing the name to just `cloudfront-functions`?

File: cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts (hunk 2 above)
Comment (paulosergio-jnr) [reply cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts:47]: I agree that OAC would be the best option here, but according to [this thread](https://github.com/aws/aws-cdk/issues/21771) we don't fully support it, as the `CloudFront Distribution` constructor always set a OAI by default when pointing to a S3 bucket as origin.
There are some work-around solutions to use OAC and remove the OAI authorization from the bucket policy, but the solution is not that "pretty", like [this one](https://github.com/aws/aws-cdk/issues/21771#issuecomment-1281190832).
What is your opinion on that? Should we implement this work-around solution or going for the not-the-best one?

Comment (ericzbeard) [reply cdk/typescript/cloudfront-functions/lib/demo-cloudfront-functions-stack.ts:47]: I guess we leave it as OAI, that's too bad CDK doesn't support OAC by default.

**Comment by ericzbeard** [comment]
> > Why did we go with cloudfront-functions as the name? Why not expand on the existing cloudfront sample?
> 
> It was just because the `-headers` suffix in the directory's name seemed to be something very specific.
> 
> WDYT of merging both examples, applying the comments above, and changing the name to just `cloudfront-functions`?

If the differences are not too significant I'm ok with that. We could also merge with the `cloudfront` app and just have one

**Comment by paulosergio-jnr** [comment]
> > > Why did we go with cloudfront-functions as the name? Why not expand on the existing cloudfront sample?
> > 
> > 
> > It was just because the `-headers` suffix in the directory's name seemed to be something very specific.
> > WDYT of merging both examples, applying the comments above, and changing the name to just `cloudfront-functions`?
> 
> If the differences are not too significant I'm ok with that. We could also merge with the `cloudfront` app and just have one

After evaluating the three different examples regarding CloudFront, I merged both that are focused on deploying CloudFront Functions as `cloudfront-functions` and let the `cloudfront` as is given these points:

- `cloudfront` example deploys the cloudfront distribution in a simpler way compared to the other two.
- `cloudfront` example has a considerable more complex comment to create the cloudfront distribution.
- `cloudfront` example is using pre-created resources, as the other two examples are creating the whole environment from scratch.

What do you think @ericzbeard ?

**Review by ericzbeard — APPROVED** [review]

**Pull request merged**